albatross/models/ransac_gp.h

```diff
@@ -0,0 +1,182 @@
+     * defining `fitter`, `outlier_metric` and `model_metric` functions
+     * that can perform each operation efficiently for Gaussian processes.
+     */
+
+    Eigen::MatrixXd cov =
+        symmetric_covariance(this->covariance_function_, features);
+
+    if (max_iterations < 1) {
+      return fit_from_covariance(features, cov, targets);
+    }
+
+    struct FitAndIndices {
+      FitType model_fit;
+      Indexer fit_indices;
+    };
+
+    std::function<FitAndIndices(const Indexer &)> fitter =
+        [&](const Indexer &indexer) {
+          const auto train_features = subset(indexer, features);
+          const auto train_targets = subset(indexer, targets);
+          auto train_cov = symmetric_subset(indexer, cov);
+          const FitAndIndices fit_and_indices = {
+              fit_from_covariance(train_features, train_cov, train_targets),
+              indexer};
+          return fit_and_indices;
+        };
+
+    std::function<double(const Indexer &, const FitAndIndices &)>
+        outlier_metric = [&](const Indexer &test_indices,
+                             const FitAndIndices &fit_and_indices) {
+          const auto cross_cov =
+              subset(fit_and_indices.fit_indices, test_indices, cov);
+          const auto pred_cov = symmetric_subset(test_indices, cov);
+          const auto pred = predict_from_covariance_and_fit(
+              cross_cov, pred_cov, fit_and_indices.model_fit);
+          const auto target = subset(test_indices, targets);
+          double metric_value = metric(pred, target);
+          return metric_value;
+        };
+
+    std::function<double(const Indexer &)> model_metric = [&](
+        const Indexer &inliers) {
+      RegressionDataset<FeatureType> inlier_dataset(subset(inliers, features),
+                                                    subset(inliers, targets));
+      const FoldIndexer inlier_loo = leave_one_out_indexer(inlier_dataset);
+
+      BaseModel model;
+      model.set_params(this->get_params());
+      double mean_metric = cross_validated_scores<double>(
+                               metric, inlier_dataset, inlier_loo, &model)
+                               .mean();
+      return mean_metric;
+    };
+
+    // Now that the ransac functions are defined we can actually perform
+    // RANSAC, then return the subsequent outlier free fit.
+    const RegressionDataset<FeatureType> dataset(features, targets);
+    const auto loo_indexer = leave_one_out_indexer(dataset);
+
+    auto inliers = ransac<FitAndIndices>(
+        fitter, outlier_metric, model_metric, map_values(loo_indexer),
+        threshold, min_features, min_inliers, max_iterations);
+
+    const auto inlier_features = subset(inliers, features);
+    const auto inlier_targets = subset(inliers, targets);
+    const auto inlier_cov = symmetric_subset(inliers, cov);
+
+    return fit_from_covariance(inlier_features, inlier_cov, inlier_targets);
+  }
+
+  double threshold;
+  std::size_t min_inliers;
```

Comment:
yeah.  I'll likely end up throwing out this entire class :) but a config struct would probably be useful in that case as well.

Reply:
Actually, it's not clear to me where to put the config struct.  Assuming the config looked like this ...
```
struct RansacConfig {
  double inlier_threshold;
  std::size_t min_inliers;
  std::size_t min_features;
  std::size_t max_iterations;
};
```
Do you think it should

1. Live in `outlier.h` (so at the low level interface to ransac).  In this case I'd either want to change the function signatures on all the ransac functions to look like:
```
ransac(const RegressionDataset<FeatureType> &dataset,
       const FoldIndexer &fold_indexer, RegressionModel<FeatureType> *model,
       EvaluationMetric<PredictType> &metric, RansacConfig &config)
```
or maintain multiple function signatures one that takes a config one that takes raw arguments but that both do the same thing.

2. live in `ransac_gp.h` In which case we'd have the same decision about whether there should be a single constructor of `RansacGaussianProcess` or multiple that all end up doing the same thing.

3. live inside `RansacGaussianProcess` which would mostly just add as a conceptual aid which would tell people looking at the code "these class members deal with configuration of the ransac algorithm not the gaussian process".

Personally I'd vote for 3 mostly because it's the easiest :) but also because I think I prefer flat arguments,
```
ransac(fitter, outlier_metric, model_metric, inlier_threshold, min_inliers, min_features, max_iterations);
```
over nested arguments
```
RansacConfig ransac_config(inlier_threshold, min_inliers, min_features, max_iterations);
ransac(fitter, outlier_metric, model_metric, ransac_config);
```
but I'm open to any of those three options.

Reply:
decided offline to just leave any of the `RansacConfig` logic to anything that wants to use the `ransac` function.  Ie, that logic will not live in `albatross`.